src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -728,8 +753,16 @@ void ZooKeeper::receiveEvent()
     ZooKeeperResponsePtr response;
     UInt64 elapsed_ms = 0;
 
-    if (unlikely(recv_inject_fault) && recv_inject_fault.value()(thread_local_rng))
-        throw Exception(Error::ZSESSIONEXPIRED, "Session expired (fault injected on recv)");
+    if (unlikely(inject_fault))
+    {
+        if (inject_mutex.try_lock_shared())
```

Comment:
I'm going to implement a lock_guard because I really dislike having to manage the mutex directly.